### DISCUSSION TITLE
Need steps to compile openssl-debian-openssl-3.0.11-1_deb12u2 with FIPS complaint openssl-3.0.8

### DISCUSSION THREAD

**Discussion by debananda-pal** [openssl/openssl#23154] [Q&A]
Hi,

Kindly let me know the steps to compile openssl-debian-openssl-3.0.11-1_deb12u2 with FIPS complaint openssl-3.0.8.
I am not sure where --with-fipsdir path will be pointing to.

Regards,
D Pal

**Comment by jamuir** [comment]
You may be interested in the following presentation on the OpenSSL FIPS provider by Matt Caswell:

https://youtu.be/geAtEXbHaFg

**Reply by debananda-pal** [reply]
Thank you Jamuir.

**Comment by debananda-pal** [comment]
@mattcaswell, Hi Matt,

We are trying to upgrade OpenSSL from Buster to Bookworm
Need your help to understand below doubts.

Is the openssl-debian-openssl-3.0.11-1_deb12u2 FIPS compliant ?
a. If the above Question answer is “No”, can I use openssl-3.0.8 ?
b. If so, please let me know the changes need to add in openssl-3.0.11 (debian/control, debian/rules and other necessary files) to use openssl-3.0.8 as FIPS module.

Regards,
D Pal

**Reply by levitte** [reply]
The way I would do it is this:

1. install Debian's openssl package 3.0.11-1 the normal way you install packages.
   *It does not include the FIPS module!*
2. download the pristine openssl 3.0.8 source from OpenSSL.
3. Configure openssl-3.0.8 to build the FIPS module, and to install in the same locations as the Debian package:
   ``` console
   $ PREFIX=/usr
   $ CONFIGDIR=$(openssl info -configdir)
   $ LIBDIR=$(openssl info -modulesdir | sed -e "s|^$PREFIX/||" -e 's|/ossl-modules$||')
   $ ./Configure enable-fips --prefix=$PREFIX --openssldir=$CONFIGDIR --libdir=$LIBDIR
   ```
4. Build and install the 3.0.8 FIPS module only
   ``` console
   $ make
   $ sudo make install_fips
   ```
5. Edit `/etc/ssl/openssl.cnf` to uncomment one line:
   ``` diff
   --- /etc/ssl/openssl.cnf~	2023-11-25 21:35:59.000000000 +0100
   +++ /etc/ssl/openssl.cnf	2023-12-28 16:54:29.766149805 +0100
   @@ -48,7 +48,7 @@
    # fips provider. It contains a named section e.g. [fips_sect] which is
    # referenced from the [provider_sect] below.
    # Refer to the OpenSSL security policy for more information.
   -# .include fipsmodule.cnf
   +.include fipsmodule.cnf
    
    [openssl_init]
    # providers = provider_sect
   ```
6. Move and link `fipsmodule.cnf` to fit how the Debian package does things:
   ``` console
   $ sudo mv /usr/lib/ssl/fipsmodule.cnf /etc/ssl
   $ sudo ln -s /etc/ssl/fipsmodule.cnf /usr/lib/ssl
   ```

I *think* this is complete.  However, *I have not tested this*...  I might do that next week, when I'll have more time to focus.

**Reply by debananda-pal** [reply]
Thank you very much levitte. I will try these steps.

**Reply by debananda-pal** [reply]
Hi, Not able to find  openssl-3.0.8 debian source tar file. If anyone can help me to get the link, will be useful.

**Reply by levitte** [reply]
When I said "download the pristine openssl 3.0.8 source from OpenSSL.", I meant https://www.openssl.org/source/old/3.0/openssl-3.0.8.tar.gz

Sorry for being unclear

**Reply by debananda-pal** [reply]
Thank you Levitte, but not able to Configure for target debian-amd64. There may be a debian directory with patches.

**Reply by mattcaswell** [reply]
> but not able to Configure for target debian-amd64.

If you followed @levitte's instructions above exactly you shouldn't be using the `debian-amd64` target at all, i.e. In @levitte's steps there is no target specified on the Configure line:

````
$ PREFIX=/usr
$ CONFIGDIR=$(openssl info -configdir)
$ LIBDIR=$(openssl info -modulesdir | sed -e "s|^$PREFIX/||" -e 's|/ossl-modules$||')
$ ./Configure enable-fips --prefix=$PREFIX --openssldir=$CONFIGDIR --libdir=$LIBDIR
````
Please show what you did and what error you encountered.

**Reply by debananda-pal** [reply]
Hi Matt, sorry for replying late.
I have resolved the debian-amd64 target issue after applying the target based patch.

For your kind information, we are not directly installing openssl 3.0.11.
We have a build procedure for many components, after building all components, the final binary will be installed on target.
Moreover, we use fixed path for --prefix, --openssldir as below
--prefix=/usr --openssldir=/usr/lib/ssl

Here I have a doubt which providers to activate on 3.0.11 and 3.0.8, as we will take 3.0.11 version as base and use 3.0.8 for FIPS part.

**Reply by levitte** [reply]
Technically, you didn't really have to apply the Debian patches on the 3.0.8 source.  The source coming from OpenSSL will build correctly on Debian machinery without the debian-adm64 config target.  That being said, it's not technically wrong to use those patches either, so you're probably fine either way.

Regard what providers to activate, that depends on what you want.  The *available* providers should all be found in /usr/lib/x86_64-linux-gnu/ossl-modules/, including the FIPS provider if you configured 3.0.8 the way I showed.

**Reply by mattcaswell** [reply]
> you didn't really have to apply the Debian patches on the 3.0.8 source.

I would say you should *not* apply the debian patches if the goal is to install the fips provider. Such patches could affect the validity of the  FIPS certificate.

**Reply by debananda-pal** [reply]
Thank you Matt for your suggestion, I will remove the debian patches.

**Comment by debananda-pal** [comment]
Hi Matt, Levitte,
Please let me know which providers I have to enable for all other functionality and FIPS mode.
I am upgrading from OpenSSL 1.1.n to 3.0.11 (buster to bookworm migration).

Regards,
D Pal

**Reply by levitte** [reply]
For FIPS mode with *our* providers, you need the FIPS provider and the base provider.

For non-FIPS mode, it's really about what you need.  The providers provider implementations for different algorithms, so you need those that give you the algorithms that you need to use.
BTW, There is a kind of related discussion in #23227 (now closed, but may still be interesting)

**Comment by debananda-pal** [comment]
Thank you Levitte

**Comment by debananda-pal** [comment]
Hi All,
Without enabling any provider, if I build the openssl-debian-openssl-3.0.11-1_deb12u2.tar.gz, it is generating .deb packages.
But if I enable any provider, many test cases are getting failed. I tried below combinations of providers :
1. base + fips (from 3.0.11 )
2. only base
3. only default.

Please let me know your suggestion how to skip the tests (if it is advisable). The tests are performed on docker build environment and not on the target device.

Regards,
D Pal

**Reply by levitte** [reply]
I don't know what Debian tool you're using to build...  but as far as I know, they all end up running `dpkg-buildpackage`.  You can control its behavior with a number of environment variables, such as `DEB_BUILD_OPTIONS` (see `man dpkg-buildpackage`).

So, short story, to avoid running the test suite as part of the build, do this first:

``` shell
export DEB_BUILD_OPTIONS
DEB_BUILD_OPTIONS=nocheck
```

Regarding your combinations, I have a small comment: don't do "only base".  It only gives you encoders and decoders, not much else, and is really meant to be used together with other providers.

**Reply by levitte** [reply]
When you're done building, don't forget to *unset* `DEB_BUILD_OPTIONS`

**Reply by debananda-pal** [reply]
Thank you Levitte.
deb packages are generated for default provider.
I enabled base and fips provider with "DEB_BUILD_OPTIONS=nocheck", but getting below error :

   debian/rules override_dh_fixperms
make[2]: Entering directory '/build/repos/workspace/debian/bookworm/x86_64/build/open-source-openSSL/openssl-3.0.11'
if [ -d debian/openssl/etc/ssl/private ] ; then \
        chmod 700 debian/openssl/etc/ssl/private ; \
fi
dh_fixperms -a -X etc/ssl/private
make[2]: Leaving directory '/build/repos/workspace/debian/bookworm/x86_64/build/open-source-openSSL/openssl-3.0.11'
   dh_missing
**dh_missing: warning: usr/lib/ssl/fipsmodule.cnf exists in debian/tmp but is not installed to anywhere**
dh_missing: error: missing files, aborting

I am able to find fipsmodule.cnf in below location of build directory :
openSSL$ find . -name fipsmodule.cnf | xargs ls -l
-rw-r--r-- 1 debananda.pal ncore 177 Jan 10 00:51 ./openssl-3.0.11/build_shared/providers/fipsmodule.cnf
-rw-r--r-- 1 debananda.pal ncore 177 Jan 10 00:40 ./openssl-3.0.11/build_static/providers/fipsmodule.cnf
-rw-r--r-- 1 debananda.pal ncore 177 Jan 10 00:56 ./openssl-3.0.11/debian/tmp/usr/lib/ssl/fipsmodule.cnf
-rw-r--r-- 1 debananda.pal ncore 235 Jan 10 00:35 ./openssl-3.0.11/test/recipes/30-test_defltfips/fipsmodule.cnf

Please let me know if I am missing anything.

Thanks,
D Pal

**Reply by debananda-pal** [reply]
Hi Matt, Levitte,
We are using below command to compile and getting above error after enabling fips mode with enable-fips option in debian/rules file
fakeroot-tcp debian/rules binary

Regards,
D Pal

**Reply by bnaigaonkar** [reply]
@debananda-pal add "usr/lib/ssl/fipsmodule.cnf" in openssl.install and execute dpkg-buildpackage -B -uc -us
Hope it will help.

As per the logs, it seems openssl compiled successfully and moved compiled files in **debian/tmp/** folder, so you can use it.
otherwise add above fix and rebuild package, this will resolve the issue and you will get a deb packages for openssl & libssl.

**Reply by debananda-pal** [reply]
Thank you bnaigaonkar.
